File: gtci/10-level-order-traversal/python3/0-level-order-traversal.py
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def getLevelOrder(self, root):
        if root is None:
            return []

        traversal = []

        queue = deque()
        queue.append(root)

        while queue:
            level_size = len(queue)

            for _ in range(level_size):
                node = queue.popleft()
                traversal.append(node.val)

                if node.left is not None:
                    queue.append(node.left)

                if node.right is not None:
                    queue.append(node.right)

        return traversal


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(4, TreeNode(5, TreeNode(5), TreeNode(7)), TreeNode(10))
    assert sol.getLevelOrder(root) == [4, 5, 10, 5, 7]

    root = TreeNode(9)
    assert sol.getLevelOrder(root) == [9]

    print("All test cases passed.")

File: gtci/10-level-order-traversal/cpp/0-level-order-traversal.cpp
#include <cassert>
#include <iostream>
#include <queue>

#include "tree_utils.h"

class Solution {
    public:
    std::vector<int> getLevelOrderTraversal(TreeNode* root) {
        if (root == nullptr)
            return {};

        std::vector<int> traversal;
        
        std::queue<TreeNode*> queue;
        queue.push(root);

        while (!queue.empty()) {
            const size_t levelSize = queue.size();

            for (size_t i = 0; i < levelSize; i++) {
                TreeNode* node = queue.front();
                queue.pop();

                traversal.push_back(node->val);

                if (node->left != nullptr)
                    queue.push(node->left);
                
                if (node->right != nullptr)
                    queue.push(node->right);
            }
        }

        return traversal;
    }
};

int main() {
    Solution sol = Solution();

    // Test case 1: Tree with multiple levels
    TreeNode* root1 = new TreeNode(4);
    root1->left = new TreeNode(5);
    root1->right = new TreeNode(10);
    root1->left->left = new TreeNode(5);
    root1->left->right = new TreeNode(7);
    std::vector<int> expected1 = {4, 5, 10, 5, 7};
    assert(sol.getLevelOrderTraversal(root1) == expected1);

    // Test case 2: Single node tree
    TreeNode* root2 = new TreeNode(9);
    std::vector<int> expected2 = {9};
    assert(sol.getLevelOrderTraversal(root2) == expected2);

    std::cout << "All test cases passed." << std::endl;
    
    // Clean up memory
    cleanupTree(root1);
    cleanupTree(root2);

    return 0;
}
